src/lib/cloud-manager-hoc.jsx

```diff
@@ -0,0 +1,98 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import {connect} from 'react-redux';
+
+import VM from 'scratch-vm';
+import CloudProvider from '../lib/cloud-provider';
+
+import {
+    getIsShowingWithId
+} from '../reducers/project-state';
+
+/*
+ * Higher Order Component to manage the connection to the cloud dserver.
+ * @param {React.Component} WrappedComponent component to manage VM events for
+ * @returns {React.Component} connected component with vm events bound to redux
+ */
+const cloudManagerHOC = function (WrappedComponent) {
+    class CloudManager extends React.Component {
+        constructor (props) {
+            super(props);
+            this.cloudProvider = null;
+            this.canUseCloud = false;
+        }
+        componentDidUpdate (prevProps) {
+            // If any of the following are not true, do not open a cloud connection
+            const canUseCloud = this.props.cloudHost && this.props.username && this.props.isShowingWithId;
+
+            if (!canUseCloud) return;
+
+            // TODO add scratcher check
+            this.canUseCloud = canUseCloud;
+
+            if ((this.props.username !== prevProps.username) ||
+                (this.props.projectId !== prevProps.projectId) ||
+                (!this.cloudProvider) ||
+                (this.cloudProvider && !this.cloudProvider.connection)) {
+                // TODO need to add provisions for viewing someone
+                // else's project in editor mode
+                this.cloudProvider = new CloudProvider(
+                    this.props.cloudHost,
+                    this.props.vm,
+                    this.props.username,
+                    this.props.projectId);
+                this.props.vm.setCloudProvider(this.cloudProvider);
+            }
+        }
+        render () {
+            const {
+                /* eslint-disable no-unused-vars */
+                cloudHost,
+                projectId,
+                username,
+                isShowingWithId,
+                /* eslint-enable no-unused-vars */
+                vm,
+                ...componentProps
+            } = this.props;
+            return (
+                <WrappedComponent
+                    canUseCloud={this.canUseCloud}
```

Comment:
It would be better to compute the value here instead of relying on a pre-computed value. This will ensure that the value of `canUseCloud` is always up-to-date given the props, whereas setting it in componentDidUpdate seems like it will always be one step behind.

Reply:
@ray, yeah I think that makes sense. This is somewhat temporary because I needed to be able to control when/when not to show the cloud variable option in the variable modal in this set of PRs. I will probably move some of this logic into www when I add more `should-we-connect??` logic, but we should maybe talk offline about what's best to have where.